### DISCUSSION TITLE
Support for default types in TypedDict

### DISCUSSION THREAD

**Discussion by fellnerse** [python/typing#1165] [General]
See here: https://github.com/python/mypy/issues/6131
What do you think?

**Comment by erictraut** [comment]
When you say "default types", am I correct in assuming that you want a way to represent that "the TypedDict is allowed to contain additional keys beyond what is specified, but they must be of this type"?

There was a recent discussion about this in the typing sig related to this topic:
https://mail.python.org/archives/list/typing-sig@python.org/thread/66RITIHDQHVTUMJHH2ORSNWZ6DOPM367/#AIBYTWKQHHVP2EYHZULI3MK4M5MB4VZA

**Reply by adam-grant-hendry** [reply]
FYI, I'm personally interested in default _arguments_. I think it would be worthwhile to separate concerns (i.e. adding support for default _types_ as well as support for default _arguments_). To that end, I have created a separate issue [#12980](https://github.com/python/mypy/issues/12980).

**Reply by erictraut** [reply]
By "default arguments", do you mean that you want the `TypedDict` constructor to provide arguments for missing keyword parameters at runtime? The problem is that `TypedDict` isn't a runtime class. It's a structural overlay on top of `dict`.

**Reply by adam-grant-hendry** [reply]
> By "default arguments", do you mean that you want the TypedDict constructor to provide arguments for missing keyword parameters at runtime? The problem is that TypedDict isn't a runtime class. It's a structural overlay on top of dict.

I want to be able to pass default arguments to class variables in `TypedDict` like so:

```python
class DefaultArgs(TypedDict):
    arg1: int = 1
    arg2: bool = True
```

The problem is there is no way currently to unpack a `TypedDict` and specify default arguments. The closest I've reached is:

```python
from __future__ import annotations

from typing import Annotated

class DefaultArgs(TypedDict):
    arg1: Annotated[int, 1]
    arg2: Annotated[bool, True]
```

From `PEP 593`:

> Specifically, a type T can be annotated with metadata x via the typehint Annotated[T, x]. This metadata can be used for either static analysis or at runtime.

and the `Consuming annotation` section

> Ultimately, the responsibility of how to interpret the annotations (if at all) is the responsibility of the tool or library encountering the Annotated type.

Is `Unpack` able to parse `Annotated` types specified in a `TypeDict` and use them to create default arguments?

**Reply by erictraut** [reply]
`TypedDict` doesn't work this way. As I mentioned above, `TypedDict` is a structural overlay on top of a `dict` runtime class. It describes the key names and associated value types expected in the dict. It therefore doesn't make sense to talk of "default arguments" in the context of a `TypedDict`. There is no way to "construct" a `TypedDict`. You need to construct a `dict` that conforms to a `TypedDict` definition. This is typically done through the use of a literal dictionary expression like `{'arg1': 1, 'arg2': True}`.

`Annotated` also doesn't really make sense here. `Annotated` is used to provide annotation metadata unrelated to type checking. The typing standard doesn't specify any meaning for metadata passed to `Annotated`. The standard provides meaning only for the primary type (the first argument to `Annotated`). The remaining arguments are ignored by standards-compliant static type checkers like mypy, pyre, and pyright. Other domain-specific tools or runtime libraries may choose to give meaning to these additional arguments, but no standards cover this.

**Reply by adam-grant-hendry** [reply]
> `TypedDict` doesn't work this way.

I'm aware. I know how `TypedDict` currently works. This is a request to change its behavior. See [TypingDict Default Arguments #12980](https://github.com/python/mypy/issues/12980).

> You need to construct a dict that conforms to a TypedDict definition. This is typically done through the use of a literal dictionary expression like {'arg1': 1, 'arg2': True}

If I had:

```python
@staticmethod
def keyClick(*args, **kwargs):
    qt_api.QtTest.QTest.keyClick(*args, **kwargs)
```

where the C++ signatures for the wrapped method are

```python
void keyClick(QWidget *widget, Qt::Key key, Qt::KeyboardModifiers modifier = Qt::NoModifier, int delay = -1)
void keyClick(QWidget *widget, char key, Qt::KeyboardModifiers modifier = Qt::NoModifier, int delay = -1)
void keyClick(QWindow *window, Qt::Key key, Qt::KeyboardModifiers modifier = Qt::NoModifier, int delay = -1)
void keyClick(QWindow *window, char key, Qt::KeyboardModifiers modifier = Qt::NoModifier, int delay = -1)
```

and a user entered something like

```python
from qtpy import QtCore, QtWidgets

widget = QtWidgets.QToolButton()

qtbot.keyClick(widget, QtCore.Qt.LeftButton)  # qtbot is a fixture from pytest-qt
```

and the `TypedDict` was

```python
class KeyClickArgs(TypedDict):
    widget: QtWidgets.QWidget | QtGui.QWindow
    key: QtCore.Qt.Key | str
    modifier: QtCore.Qt.KeyboardModifiers
    delay: int
```

then `mypy` will complain that `modifier` and `delay` aren't specified. We would be forcing the user to enter the default arguments anyway, which defeats the purpose of having default arguments in the first place.

Hence, using a literal dictionary expression is not a viable alternative for wrapper functions.

> `Annotated` also doesn't really make sense here. 

Yes, I understand. I mentioned this [here](https://github.com/python/mypy/issues/12980#issuecomment-1154605393) where I said:

> I think currently default arguments might be possible using the Annotated type, but I don't think that was necessarily its intended purpose as specified in [PEP 593](https://peps.python.org/pep-0593/) so it doesn't feel pythonic to use

I used `Annotated` because it is currently the only structure I know I can use in a `TypedDict` that `mypy` won't complain about and has compatibility for runtime as well as static analysis purposes. Even if a same behavior as `Annotated` could be used, I would strongly argue the name be changed to something that makes more sense when used in `TypedDict`s, like `Default` or similar.

---

Rather than adding a literal as a default argument, can `TypedDict` be modified to adopt the ellipses to specify that a key contains a default type, like so?:

```python
class KeyClickArgs(TypedDict):
    widget: QtWidgets.QWidget | QtGui.QWindow
    key: QtCore.Qt.Key | str
    modifier: QtCore.Qt.KeyboardModifiers = ...
    delay: int = ...
```

**Reply by adam-grant-hendry** [reply]
Or @erictraut , perhaps this is where `Required` and `NotRequired` should be used? Is that the case?

**Reply by erictraut** [reply]
> I'm aware. I know how TypedDict currently works. This is a request to change its behavior.

What you're proposing has no relation to the existing `TypedDict`. "Default arguments" don't make sense for a structural type definition. `TypedDict` is not the solution you're looking for. 

Let's back up and look at the problem you're trying to solve. If I understand you correctly, you want to provide an annotated signature for the `keyClick` method. Here's how that would look:

```python
@staticmethod
def keyClick(widget: QWidget, key: Qt.Key | str, modifier: Qt.KeyboardModifiers = ..., delay: int  = ...) -> None: ...
```

Does that solve your problem?

**Reply by adam-grant-hendry** [reply]
> Does that solve your problem?

<strike>Unfortunately, I don't have the power to change the 3rd party library.</strike> This is the stub, sorry. I see now.

Yes, I believe that would solve my problem. However, it doesn't feel true to the implementation (which I cannot change and uses `*args` and `**kwargs`). Instead, could I do:

```python
class KeyClickArgs(TypedDict):
    widget: Required[QtWidgets.QWidget | QtGui.QWindow]
    key: Required[QtCore.Qt.Key | str]
    modifier: NotRequired[QtCore.Qt.KeyboardModifiers]
    delay: NotRequired[int]
```

and `Unpack` that?

**Reply by adam-grant-hendry** [reply]
@erictraut I think perhaps what I am looking for is precisely what you're aiming for in [PEP 692](https://github.com/python/peps/pull/2620/files). You're entirely correct that

```python
@staticmethod
def keyClick(widget: QWidget, key: Qt.Key | str, modifier: Qt.KeyboardModifiers = ..., delay: int  = ...) -> None: ...
```

solves my problem, but I'm desiring to be truer to the implementation in this specific instance and being able to type hint ``**kwargs`` with a variadic TypeDict like so

```python
class KeyClickArgs(TypedDict):
    widget: Required[QtWidgets.QWidget | QtGui.QWindow]
    key: Required[QtCore.Qt.Key | str]
    modifier: NotRequired[QtCore.Qt.KeyboardModifiers]
    delay: NotRequired[int]

@staticmethod
def keyClick(**kwargs: KeyClickArgs) -> None: ...
```

would be preferable. However, as you rightly mention in the proposal:

> However, specifying ``**kwargs`` type with a ``TypedDict`` means, as mentioned earlier, that each keyword argument specified by ``**kwargs`` is a ``TypedDict`` itself...Therefore, in order to support specifying ``kwargs`` type as a ``TypedDict`` without breaking current behaviour, a new syntax has to be introduced.

In fact, for `PyQt`/`PySide`, one can **only** use keyword arguments for default arguments. All others (e.g. `widget` and `key`) **must** be positional only. Doing otherwise gives a runtime error. (See [PYSIDE-1964](https://bugreports.qt.io/browse/PYSIDE-1964) where they explain it's not possible to give keyword arguments to all parameters).

Hence, I think I actually need something like this:

```python
class KeyClickKwargs(TypedDict):
    modifier: NotRequired[QtCore.Qt.KeyboardModifiers]  # Defaults to QtCore.Qt.NoModifier
    delay: NotRequired[int]  # Defaults to -1

@staticmethod
def keyClick(*args: *tuple[QtWidgets.QWidget | QtGui.QWindow, QtCore.Qt.Key | str], **kwargs: **KeyClickKwargs) -> None: ...
```

but I'm not sure if the syntax above is correct (I know the `**kwargs` part won't work until `PEP 692` is approved).

**Reply by erictraut** [reply]
> it doesn't feel true to the implementation

Why not? It accurately represents the set of parameters and their types, doesn't it?

> I'm desiring to be truer to the implementation

You are being true to the implementation by spelling out the parameters in the signature. You need to use the right tools for the job though. Using a `TypedDict` and `**kwargs` annotations are the wrong tools for this job. It's like saying "I have a hammer, but I need a mode of transportation, so can I add wheels to my hammer?" ;). In this case, a normal function signature in the type stub is the right tool for the job.

> In fact, for PyQt/PySide, one can only use keyword arguments for default arguments. All others (e.g. widget and key) must be positional only. 

You can use positional-only and keyword-only separators in a function signature to specify this.

```python
# a is positional-only
# b is positional or keyword
# c is keyword-only
def foo(a: int, /, b: str, *, c: float) -> None: ...
```

See [PEP 570](https://peps.python.org/pep-0570/) for more details.

**Reply by adam-grant-hendry** [reply]
+1 on `PEP 570`. I didn't know about that.

> Why not? It accurately represents the set of parameters and their types, doesn't it?

Yes and no. While those are the true parameters and types, the library author chose to use `*args` and `**kwargs`. If I could change the implementation I would because I believe explicit is better than implicit, but with

```python
class KeyClickKwargs(TypedDict):
    modifier: NotRequired[QtCore.Qt.KeyboardModifiers]  # Defaults to QtCore.Qt.NoModifier
    delay: NotRequired[int]  # Defaults to -1

@staticmethod
def keyClick(*args: *tuple[QtWidgets.QWidget | QtGui.QWindow, QtCore.Qt.Key | str], **kwargs: **KeyClickKwargs) -> None: ...
```

There feels a direct link between how the author wrote the signature

```python
def keyClick(*args, **kwargs)
```

as opposed to:

```python
def keyClick(widget, key, modifier, delay)
```

> Using a TypedDict and **kwargs annotations are the wrong tools for this job.

Perhaps currently, but isn't that precisely what [PEP 692](https://github.com/python/peps/pull/2620/files) is aiming for?

**Reply by erictraut** [reply]
I don't know why the author chose to use `*args` and `**kwargs`, but that's an implementation detail. The important thing is that you are accurately and completely specifying the interface contract for the function. The Python type system has well-established tools for doing that.

> Perhaps currently, but isn't that precisely what PEP 692 is aiming for?

No, not really. PEP 692 augments the current type system to handle some edge cases that are not possible today. In particular, it is designed to handle the case where a function is designed to be used with a dictionary passed to `**kwargs`. It allows you to overlay a structural type on that dictionary to provide more precise type checking. The `keyClick` function is not designed to be called with a dictionary. It is designed to be called with a combination of positional and keyword arguments, and the existing mechanisms in the Python type system work well to describe its signature.

**Reply by adam-grant-hendry** [reply]
> The keyClick function is not designed to be called with a dictionary.

Forgive my ignorance, but if the wrapper function is defined using *args and **kwargs, doesn’t that mean it’s designed to be called with a dictionary? The original might not be defined that way, but the wrapper is.

I’m not sure I understand the distinction you’re making? Any function defined with **kwargs should be callable either explicitly with a dictionary or keyword arguments.

Can you show me an example of where a function defined with **kwargs _can’t_ be called using keyword arguments and must use a dictionary? That would help my understanding.

**Reply by adam-grant-hendry** [reply]
Better yet, let me ask the question another way. Assuming PEP 692 were already approved and implemented, why _shouldn’t_ I type hint keyClick this way:


class KeyClickKwargs(TypedDict):
    modifier: NotRequired[QtCore.Qt.KeyboardModifiers]  # Defaults to QtCore.Qt.NoModifier
    delay: NotRequired[int]  # Defaults to -1

@staticmethod
def keyClick(*args: *tuple[QtWidgets.QWidget | QtGui.QWindow, QtCore.Qt.Key | str], **kwargs: **KeyClickKwargs) -> None: ...

**Reply by erictraut** [reply]
The `keyClick` function is designed to be called with either positional or keyword arguments. The [documentation](https://doc.bccnsoft.com/docs/PyQt4/qtest.html#keyClick) is a better indication of its intended usage than the implementation.

> why shouldn’t I type hint keyClick this way?

First, that doesn't correctly represent the set of parameters that the function accepts. For example, you wouldn't be able to pass the `widget` and `key` arguments by keyword, which is permitted by the function.

Second, language servers that offer signature help would present a signature that would be really confusing to users. You would see something like this:

![Screen Shot 2022-06-19 at 12 05 24 PM](https://user-images.githubusercontent.com/7040122/174496841-a1ae9d12-0158-4eb2-b952-95dd96f930b8.png)

Rather than this:

![Screen Shot 2022-06-19 at 12 12 30 PM](https://user-images.githubusercontent.com/7040122/174496844-89efa0d9-87b2-46f6-b411-50c2cbcf6c74.png)

You're pretty far down the rabbit hole on this one. Please use the type system and tools in their intended manner, and it will work well.

**Reply by adam-grant-hendry** [reply]
+1 for the language server. That’s a good reason.

> You're pretty far down the rabbit hole on this one. 

Yes, I understand and thank you for your help. I’m just trying to understand things.

Can you please explain what PEP 692 is for then? In what cases should this be used/is it necessary?

**Reply by adam-grant-hendry** [reply]
@erictraut Thank you again for your help (and Happy Father's Day as well).

From our conversation, I’m inclined to believe you are against PEP 692, and I understand why.

Ironically, part of its motivation states it is useful for:

> existing codebases where the need of refactoring the code in order to introduce proper type annotations may be considered not worth the effort. This in turn prevents the project from getting all of the benefits that type hinting can provide.

even though you showed the additional type hinting is a rather less than useful

**kwargs: **KwargsTypedDict

Ultimately, I see now that PEP 692 is a convenience rather than a necessity. It should be up to the purview of the coder to choose whether or not to use it.

To that end, I will recommend an additional note be added to the PEP that recommends against its usage wherever possible as it tends to abstract away useful type hinting information rather than add to it. In all cases, it would seem better to simply type out the information rather than hide it behind a TypedDict.

**Reply by erictraut** [reply]
> I’m inclined to believe you are against PEP 692

On the contrary, I'm supportive of PEP 692. In fact, I implemented support for the mechanism in pyright when Franek was working on the initial draft of the PEP.

The facility described in PEP 692 has utility, but it is applicable only for certain use cases. It is not a replacement for the existing, well-established mechanisms in the type system for describing function signatures with keyword parameters.

PEP 692 is useful in cases where an API accommodates optional keyword-only parameters when no default values are provided for absent keywords. The draft PEP shows some examples. Here's another [discussion](https://github.com/python/mypy/issues/4441) in the mypy issue tracker that provides some examples where it might be useful. Note that most functions written in Python, even if they use `**kwargs` in their implementation, do not follow this design pattern. That explains why the mechanism described in PEP 692 is a latecomer to the type system compared to other well-established facilities like specification of positional-only parameters, keyword-only parameters, and overloads. I predict PEP 692 will be used relatively infrequently, but there is clearly demand for it. I hope that it won't be abused.

@adam-grant-hendry, the fact that the draft PEP 692 created so much confusion for you leaves me a bit worried. For reference, are you relatively new to static typing in Python, or have you been using typed Python for a long time? If it confuses other Python developers to the same degree, it has the potential of being a net negative to the type system.

@franekmagiera, perhaps it's worth adding some language in the PEP (and maybe an example) that clarifies where the new mechanism should _not_ be used?

**Reply by adam-grant-hendry** [reply]
> For reference, are you relatively new to static typing in Python, or have you been using typed Python for a long time?

Not terribly new, no.I have over 15 years experience with Python and C++. The concepts of generics and variadics are not foreign to me and I know type hinting was first proposed by Guido back in 2015, so it has been around a while. However, for the sake of argument, I will consider myself relatively new for now.

From a beginner's standpoint, I think your fear that PEP 692 might be abused is warranted on at least three points (hopefully my notes might help you and others with some of the wording in the documentation):

1. I wasn't aware originally that if a function signature implementation uses `*args` and `**kwargs`

```python
@staticmethod
def keyClick(*args, **kwargs):
    # Implementation goes here...
```

that one could type hint it with the actual intended arguments

```python
@staticmethod
def keyClick(widget: QWidget, key: Qt.Key | str, modifier: Qt.KeyboardModifiers = ..., delay: int  = ...) -> None: ...
```

without using `@overload`. Furthermore, to use `@overload`, one needs to have at least 2 type signatures, so I thought I would have to figure out how to type hint using the `*args`/`**kwargs` variant anyway, so using `@overload` wouldn't have made sense for me.

I feel I read PEP 484 very well, but I couldn't gather that that was possible. Furthermore, I still get from `pylance` the linting errors:

```
Type of parameter "args" is unknown.
Type of parameter "kwargs" is unknown.
```

even if my stub is type hinted to the latter, so it would still not pass `pre-commit`.

2. For those who do want to include type hints in the source code itself, rather than in `*.pyi` stubs, (unless I'm mistaken) it does seem that one would have to refactor their code in order to type hint:

```python
@staticmethod
def keyClick(*args, **kwargs):
    # Implementation goes here...
```

would have to change to

```python
@staticmethod
def keyClick(widget: QWidget, key: Qt.Key | str, modifier: Qt.KeyboardModifiers = ..., delay: int  = ...) -> None:
    # Implementation goes here...
```

which could be a problem for 3rd party libraries if the original maintainers do not want to refactor signatures, but want contributors to develop type hints directly within the code.

3. I'm concerned about the language server hiding type information, as you've shown. This inclines me not to use it in favor of refactoring altogether.

> perhaps it's worth adding some language in the PEP (and maybe an example) that clarifies where the new mechanism should not be used?

I am fully in support of this! In addition, it would also be useful if `pylance` could output a warning or a hint if it detects that it is being used in a way that is not intended by PEP 692.

**Reply by franekmagiera** [reply]
That's an interesting discussion. I agree that in the example you've been discussing using an "explicit" signature is the best tool for the job. 

As you've noticed in point 2 that is not always possible without refactoring. I think PEP 692 would still bring some value - realistically not everyone will want to refactor, in that case I'd much rather see `**kwargs: **SomeTypedDictThatIHaveToLookUp` than `**kwargs: Any` (also seems like language servers should be able to present unpacked **kwargs in that case) - also then `kwargs` variable inside the function could be type checked (and treated as a `TypedDict`) and that would bring benefits to the maintainers of the "upstream" codebase. Point 2 of your last reply is exactly where I'm coming from.

That said, it is definitely worth including the conclusions from this thread in the draft!

**Comment by adam-grant-hendry** [comment]
> When you say "default types", am I correct in assuming that you want a way to represent that "the TypedDict is allowed to contain additional keys beyond what is specified, but they must be of this type"?

@erictraut This is correct w.r.t. @rra 's original request in [python/mypy/6131](https://github.com/python/mypy/issues/6131).

@fellnerse and I seem to be discussing default _values_.

**Comment by spacether** [comment]
One can make a dict with an additionalProperty type today in vscode per: https://stackoverflow.com/questions/73558897/in-python-how-can-i-make-a-dict-that-has-type-hints-for-know-keys-and-a-default/73559006?noredirect=1#comment129898346_73559006

**Comment by pycaw** [comment]
Currently it seems neither TypedDict+Unpack nor ParamSpec nor Protocol support (proper) forwarding of function parameter annotations. They almost do it but then fall short in differents ways. One then needs to resort to using explicit parameter classes.
It would be really great there were a way to define a set of (kw) parameters that could be reused in function definitions with hints, defaults, docs preserved.
Probably not a fair comparision, but now I value std::forward, perfect forwarding and the DRY that was/is possible to achieve with them in C++ much more.
